app/hooks/header.ts

```diff
@@ -28,93 +29,95 @@ export const useDefaultHeaderHeight = () => {
-    if (hasLargeTitle && hasSubtitle && !hasSearch) {
-        return defaultHeight + ViewConstants.LARGE_HEADER_TITLE + ViewConstants.HEADER_WITH_SUBTITLE;
-    } else if (hasLargeTitle && hasSearch) {
-        return defaultHeight + ViewConstants.LARGE_HEADER_TITLE + ViewConstants.HEADER_WITH_SEARCH_HEIGHT;
-    }
-
-    return defaultHeight + ViewConstants.LARGE_HEADER_TITLE;
+    return defaultHeight + ViewConstants.LARGE_HEADER_TITLE + ViewConstants.HEADER_WITH_SUBTITLE;
 };
 
-export const useHeaderHeight = (hasLargeTitle: boolean, hasSubtitle: boolean, hasSearch: boolean) => {
+export const useHeaderHeight = () => {
     const defaultHeight = useDefaultHeaderHeight();
-    const largeHeight = useLargeHeaderHeight(hasLargeTitle, hasSubtitle, hasSearch);
+    const largeHeight = useLargeHeaderHeight();
     return useMemo(() => {
         return {
             defaultHeight,
             largeHeight,
         };
-    }, [defaultHeight, hasSearch, largeHeight]);
+    }, [defaultHeight, largeHeight]);
 };
 
-export const useCollapsibleHeader = <T>(isLargeTitle: boolean, hasSubtitle: boolean, hasSearch: boolean) => {
+export const useCollapsibleHeader = <T>(isLargeTitle: boolean, onSnap?: (offset: number) => void) => {
     const insets = useSafeAreaInsets();
     const animatedRef = useAnimatedRef<Animated.ScrollView>();
-    const {largeHeight, defaultHeight} = useHeaderHeight(true, hasSubtitle, hasSearch);
+    const {largeHeight, defaultHeight} = useHeaderHeight();
     const scrollValue = useSharedValue(0);
+    const autoScroll = useSharedValue(false);
+    const snapping = useSharedValue(false);
+
+    const headerHeight = useDerivedValue(() => {
+        const minHeight = defaultHeight + insets.top;
+        const value = -(scrollValue?.value || 0);
+        const header = (isLargeTitle ? largeHeight : defaultHeight);
+        const height = header + value + insets.top;
+        if (height > header + (insets.top * 2)) {
+            return Math.min(height, largeHeight + insets.top + 80);
+        }
+        return Math.max(height, minHeight);
+    });
 
     function snapIfNeeded(dir: string, offset: number) {
         'worklet';
-        if (dir === 'up' && offset < defaultHeight) {
-            const diffHeight = largeHeight - defaultHeight;
-            let position = 0;
-            if (Platform.OS === 'ios') {
-                position = (diffHeight - (hasSearch ? -HEADER_SEARCH_BOTTOM_MARGIN : insets.top));
-            } else {
-                position = hasSearch ? largeHeight + HEADER_SEARCH_BOTTOM_MARGIN : diffHeight;
-            }
-            scrollTo(animatedRef, 0, position, true);
-        } else if (dir === 'down') {
-            const inset = largeHeight + (hasSearch ? HEADER_SEARCH_BOTTOM_MARGIN : 0);
-            if (offset < inset) {
-                scrollTo(animatedRef, 0, -insets.top, true);
+        if (onSnap && !snapping.value) {
+            snapping.value = true;
```

Comment:
Why the "snapping guard"? It seems to be doing some synchronous task, so it will never (in theory) execute anything else between the two calls. Am I right?